Exclude yanked releases when choosing baseline version

mgr0dzicki opened the pull request:
Closes #254

Extracts the selection of baseline version to a separate function, adds tests for it.

tonowak commented on src/baseline.rs:
```diff
@@ -340,3 +356,82 @@ fn need_retry(res: Result<(), crates_index::Error>) -> anyhow::Result<bool> {
+    fn new_crate(versions: &[Version]) -> Crate {
+        // `crates_index::Crate` cannot be created explicitly, as its field
+        // is private, so we use the fact that it can be deserialized.
+        serde_json::from_value(serde_json::json!({ "versions": versions })).unwrap()
```
From what I remember, we're trying to use `expect` (or any alternatives with custom error messages) instead of `unwrap` for better debugging messages.

mgr0dzicki replied:
Changed all `unwrap`s to `expect`s, please have a look at the messages.

tonowak commented on src/baseline.rs:
```diff
@@ -340,3 +356,82 @@ fn need_retry(res: Result<(), crates_index::Error>) -> anyhow::Result<bool> {
+    #[test]
+    fn choose_baseline_version_yanked() {
+        let crate_ = new_crate(&[
+            new_mock_version("1.2.0", false),
```
Each `new_crate` call is done by passing a list of `new_mock_version`. I wonder, wouldn't it be more concise to pass a list of tuples/structs (`version`, `is_yanked`), instead of a list of objects constructed by passing `version` and `is_yanked`?

This comment is not a request to change anything in the code -- I'm just curious whether this would be more readable/concise and I'm open for discussion.

mgr0dzicki replied:
I've done some refactoring to avoid this code duplication in tests, how do you find it now?

tonowak commented on src/baseline.rs:
```diff
@@ -340,3 +356,82 @@ fn need_retry(res: Result<(), crates_index::Error>) -> anyhow::Result<bool> {
+    }
+
+    #[test]
+    fn choose_baseline_version_yanked() {
```
This function name is misleading. It is not a yanked version that is being chosen.

mgr0dzicki replied:
I'm not really good at naming thing, what better proposition do you have?

tonowak replied:
What do you think about `baseline_choosing_logic_skips_yanked`?

tonowak commented on src/baseline.rs:
```diff
@@ -340,3 +356,82 @@ fn need_retry(res: Result<(), crates_index::Error>) -> anyhow::Result<bool> {
+    }
+
+    #[test]
+    fn choose_baseline_version_no_current() {
```
This function name is possibly misleading. From the name, I would expect the test to have a `current` version and a list of crates that doesn't have this `current` version.

mgr0dzicki replied:
And same question here, do you have some idea how can we name it better?

tonowak replied:
What do you think about `baseline_choosing_logic_without_current_picks_latest_normal`? A bit long, but I don't see any reason to make it shorter.

tonowak commented on src/baseline.rs:
```diff
@@ -242,6 +243,43 @@ fn create_rustdoc_manifest_for_crate_version(
+    } else {
+        let instance = crate_
+            .highest_normal_version()
+            .unwrap_or_else(|| crate_.highest_version())
```
I know that this code is unchanged in this PR, but I don't really understand this line -- maybe it could be commented to explain what's going on here @obi1kenobi. What is considered a "normal" version and why it is OK to choose a "not normal" (whatever is "normal") version when there isn't a normal one? Assuming that a "normal" possibly means non-yanked and non-pre-release, what would happen when there are only "not normal" versions?

obi1kenobi replied:
Your guess was correct, it's "not yanked, not pre-release": https://docs.rs/crates-index/latest/crates_index/struct.Crate.html#method.highest_normal_version

If there's no "normal" version, then our options are:
- error
- choose the best option from what's available

The choice isn't necessarily obvious, and this isn't my code. But I think the decision to try to "guess what the user wants" in this case rather than making it a hard error is a reasonable one: in the worst case, we error anyway, and in the best case we guess correctly and delight the user.

Leaving a comment is reasonable. It might just restate what the docs for the two methods on `Crate` already say, but that's alright if it saves the next reader of the code a Google search.

tonowak replied:
Can you leave a comment here (in this code) in this PR @mgr0dzicki ?

mgr0dzicki replied:
Done! Please have a look at it @tonowak

obi1kenobi commented on src/baseline.rs:
```diff
@@ -259,30 +297,8 @@ impl BaselineLoader for RegistryBaseline {
         // - Most likely the user cares about the last official release
```
This comment block seems relevant to the code block you extracted into a function. Probably move it there too?

obi1kenobi reviewed (commented):
I'm excited to have test coverage on this tricky logic! Thanks for putting it together.

It would be good to add two more test cases that check & document that when all versions are either yanked or pre-releases, cargo-semver-checks will pick the largest-numbered version regardless of whether it's been yanked or is a pre-release. (One test where it's yanked and one where it's a pre-release.)

mgr0dzicki commented:
> It would be good to add two more test cases that check & document that when all versions are either yanked or pre-releases, cargo-semver-checks will pick the largest-numbered version regardless of whether it's been yanked or is a pre-release. (One test where it's yanked and one where it's a pre-release.)

Added! They revealed that my logic of handling yanked versions was not behaving as it should (it was ignoring yanked releases when there weren't any normal one). I fixed this.

tonowak commented on src/baseline.rs:
```diff
@@ -340,3 +360,104 @@ fn need_retry(res: Result<(), crates_index::Error>) -> anyhow::Result<bool> {
+            "yanked": yanked,
+            "cksum": "00".repeat(32),
+        }))
+        .expect("Parsing JSON as crates_index::Version should not result in any errors.")
```
Something I learned from @obi1kenobi -- this message describes an invariant, but it makes more sense in an error message to describe what failed. So maybe consider writing it as `Failed to create crates_index::Version from a hand-written JSON`.

tonowak commented on src/baseline.rs:
```diff
@@ -374,64 +374,68 @@ mod tests {
+            .expect("Parsing JSON as crates_index::Crate should not result in any errors.")
+    }
+
+    fn test_choose_baseline_version(
```
This function name doesn't tell much.

tonowak commented on src/baseline.rs:
```diff
@@ -340,3 +363,113 @@ fn need_retry(res: Result<(), crates_index::Error>) -> anyhow::Result<bool> {
+                .collect(),
+        );
+        let current_version = current_version_name.map(|version_name| {
+            semver::Version::parse(version_name).expect("current_version_name should be valid")
```
Here you also have `expect`s as invariants.

tonowak commented on src/baseline.rs:
```diff
@@ -340,3 +363,113 @@ fn need_retry(res: Result<(), crates_index::Error>) -> anyhow::Result<bool> {
+            semver::Version::parse(version_name).expect("current_version_name should be valid")
+        });
+        let choosen_baseline = choose_baseline_version(&crate_, current_version.as_ref())
+            .expect("choose_baseline_version should not return any error");
```
I think it should be `chosen`, not `choosen`?

mgr0dzicki replied:
definitely ;)

tonowak commented on src/baseline.rs:
```diff
@@ -340,3 +363,105 @@ fn need_retry(res: Result<(), crates_index::Error>) -> anyhow::Result<bool> {
+            "yanked": yanked,
+            "cksum": "00".repeat(32),
+        }))
+        .expect("Failed to create crates_index::Version from a hand-written JSON.")
```
@obi1kenobi can you describe what exactly should be the format of `expect` messages? I remember that you said that they should start with a small letter and not end with a comma. But also, here it's explained that the string should describe an invariant, not what went wrong: https://stackoverflow.com/questions/66362625/why-is-rusts-expect-called-expect

obi1kenobi replied:
This is worth a read: https://doc.rust-lang.org/std/error/index.html#common-message-styles

Definitely lowercase and not ending with punctuation. For the rest, read the doc at the link above and then use your best judgment :)

mgr0dzicki replied:
Changed to lowercase without ending punctuation. I used the “expect as precondition” style everywhere, hope it's finally good enough ;)

tonowak commented on src/baseline.rs:
```diff
@@ -340,3 +363,105 @@ fn need_retry(res: Result<(), crates_index::Error>) -> anyhow::Result<bool> {
+                .expect("Failed to parse hand-written string as a valid version name.")
+        });
+        let chosen_baseline = choose_baseline_version(&crate_, current_version.as_ref())
+            .expect("choose_baseline_version should not return any error");
```
Inconsistent style with the rest of `expect`s.

obi1kenobi commented on src/baseline.rs:
```diff
@@ -242,6 +243,52 @@ fn create_rustdoc_manifest_for_crate_version(
+                )
+            })
+    } else {
+        // If there is no normal version (not yanked and not a pre-release)
+        // choosing the latest one anyway is more reasonable than throwing an
+        // error, as there is still a chance that it is what the user expects.
```
This comment isn't in the right place, and is confusing as a result. Its current placement seems to indicate that it's summarizing what's happening in the entire `else` block. But instead, the comment only applies to the `.unwrap_or_else(|| crate_.highest_version())` part.

I'd make that closure multi-line, and put the comment inside its `{` block instead. Then it's impossible to misinterpret.

mgr0dzicki replied:
Definitely good idea, thank you!

obi1kenobi reviewed (approved):
This looks good to me! @tonowak if your concerns are also resolved, let's merge?
